File: app/routers/sync.py
"""Sync endpoints — Gmail, Mercado Pago, Plaid, y Prometeo."""
from datetime import datetime
from pydantic import BaseModel
from fastapi import APIRouter, Depends, HTTPException, BackgroundTasks, UploadFile, File
from sqlalchemy.orm import Session
from ..database import get_db
from ..models import Connection, Transaction, User
from ..security import get_current_user
from ..services.dedup import find_cross_source_duplicate, mark_duplicates_and_transfers
from ..services.splits import detect_split_candidates
from ..services.classifier import classify
from ..services.alert_engine import run_alert_engine
from ..services.plaid_sync import plaid_client
from ..services.prometeo_api import get_prometeo_client

router = APIRouter(prefix="/api/sync", tags=["sync"])


async def _save_transaction_item(item: dict, user_id: int, db: Session) -> bool:
    """Save one transaction. Returns True if saved, False if duplicate."""
    raw_ref = str(item.get("id") or item.get("raw_reference", ""))
    source = item.get("source", "")

    if raw_ref:
        exists = db.query(Transaction).filter(
            Transaction.user_id == user_id,
            Transaction.raw_reference == raw_ref,
            Transaction.source == source,
        ).first()
        if exists:
            return False

    dup = find_cross_source_duplicate(db, user_id, item)

    tx_type = item.get("tx_type", "expense")
    payment_method = item.get("payment_method", "") or ""

    if tx_type == "income":
        tx = Transaction(
            user_id=user_id,
            source=source,
            tx_type="income",
            amount=item["amount"],
            currency=item.get("currency", "ARS"),
            date=item["date"],
            month=item.get("month", item["date"][:7]),
            merchant=item.get("merchant", ""),
            provider=item.get("provider", ""),
            category="ingreso",
            subcategory="",
            status="confirmed",
            confidence=1.0,
            payment_method=payment_method,
            needs_review=item.get("needs_review", False),
            raw_reference=raw_ref,
            is_duplicate=dup is not None,
        )
    else:
        result = await classify(
            item.get("merchant", ""),
            item.get("amount", 0),
            source,
            db,
            user_id=user_id,
        )
        tx = Transaction(
            user_id=user_id,
            source=source,
            tx_type="expense",
            amount=item["amount"],
            currency=item.get("currency", "ARS"),
            date=item["date"],
            month=item.get("month", item["date"][:7]),
            merchant=item.get("merchant", ""),
            provider=item.get("provider", ""),
            category=result["category"],
            subcategory=result.get("subcategory", ""),
            status="confirmed",
            confidence=result.get("confidence", 0.7),
            rule_used=result.get("rule_used"),
            ai_reason=result.get("reason"),
            payment_method=payment_method,
            needs_review=item.get("needs_review", False) or result.get("needs_review", False),
            raw_reference=raw_ref,
            is_duplicate=dup is not None,
        )

    db.add(tx)
    db.commit()
    return True


async def _save_transactions(items: list[dict], user_id: int, db: Session) -> int:
    """Save a list of transaction items. Returns count of newly saved."""
    saved = 0
    for item in items:
        if await _save_transaction_item(item, user_id=user_id, db=db):
            saved += 1
    return saved


@router.post("/gmail")
async def sync_gmail(background_tasks: BackgroundTasks, db: Session = Depends(get_db), user: User = Depends(get_current_user)):
    conn = db.query(Connection).filter_by(user_id=user.id, provider="gmail").first()
    if not conn or conn.status != "connected" or not conn.access_token:
        raise HTTPException(400, "Gmail no conectado. Conectalo desde Configuración.")

    from ..services.gmail import fetch_payment_emails
    try:
        items = await fetch_payment_emails(conn.access_token)
    except Exception as e:
        raise HTTPException(502, f"Error al leer Gmail: {str(e)}")

    saved = await _save_transactions(items, user_id=user.id, db=db)

    conn.last_sync = datetime.utcnow()
    db.commit()
    flagged = mark_duplicates_and_transfers(db, user.id)
    splits = detect_split_candidates(db, user.id)

    background_tasks.add_task(run_alert_engine, user.id, db)

    return {"ok": True, "fetched": len(items), "saved": saved, "flagged": flagged, "split_suggestions": splits}


@router.post("/mp")
async def sync_mp(background_tasks: BackgroundTasks, db: Session = Depends(get_db), user: User = Depends(get_current_user)):
    conn = db.query(Connection).filter_by(user_id=user.id, provider="mercadopago").first()
    if not conn or conn.status != "connected" or not conn.access_token:
        raise HTTPException(400, "Mercado Pago no conectado. Conectalo desde Configuración.")

    from ..services.mercadopago import fetch_movements
    try:
        items = await fetch_movements(conn.access_token)
    except Exception as e:
        raise HTTPException(502, f"Error al leer Mercado Pago: {str(e)}")

    saved = await _save_transactions(items, user_id=user.id, db=db)

    conn.last_sync = datetime.utcnow()
    db.commit()
    flagged = mark_duplicates_and_transfers(db, user.id)
    splits = detect_split_candidates(db, user.id)

    background_tasks.add_task(run_alert_engine, user.id, db)

    return {"ok": True, "fetched": len(items), "saved": saved, "flagged": flagged, "split_suggestions": splits}


@router.post("/csv")
async def sync_csv(
    background_tasks: BackgroundTasks,
    file: UploadFile = File(...),
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
):
    """Importa el estado de cuenta de MP (CSV). Cubre compras con tarjeta
    que la API pública no expone. Deduplica contra lo ya sincronizado."""
    if file.size and file.size > 5_000_000:
        raise HTTPException(413, "Archivo muy grande (máximo 5 MB)")
    content = await file.read()
    from ..services.csv_import import parse_mp_csv
    try:
        items = parse_mp_csv(content)
    except ValueError as e:
        raise HTTPException(400, str(e))

    saved = await _save_transactions(items, user_id=user.id, db=db)
    flagged = mark_duplicates_and_transfers(db, user.id)
    splits = detect_split_candidates(db, user.id)
    background_tasks.add_task(run_alert_engine, user.id, db)
    return {"ok": True, "fetched": len(items), "saved": saved,
            "flagged": flagged, "split_suggestions": splits}


@router.get("/status")
def sync_status(db: Session = Depends(get_db), user: User = Depends(get_current_user)):
    conns = db.query(Connection).filter_by(user_id=user.id).all()
    return {
        c.provider: {
            "status": c.status,
            "last_sync": c.last_sync.isoformat() if c.last_sync else None,
        }
        for c in conns
    }


# === PLAID ENDPOINTS ===

@router.post("/plaid/link_token")
async def get_plaid_link_token(user: User = Depends(get_current_user)):
    """Obtener link token para conectar con Plaid"""
    link_token = await plaid_client.create_link_token(user.id, user.email)
    if not link_token:
        raise HTTPException(status_code=400, detail="Error creando link token de Plaid")
    return {"link_token": link_token}


@router.post("/plaid/exchange_token")
async def exchange_plaid_token(
    public_token: str,
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Intercambiar public token por access token y guardar conexión"""
    access_token = await plaid_client.exchange_token(public_token)
    if not access_token:
        raise HTTPException(status_code=400, detail="Error intercambiando token")

    # Guardar o actualizar conexión en DB
    connection = db.query(Connection).filter_by(
        user_id=user.id,
        provider='plaid'
    ).first()

    if connection:
        connection.access_token = access_token
        connection.status = 'connected'
        connection.last_sync = datetime.utcnow()
    else:
        connection = Connection(
            user_id=user.id,
            provider='plaid',
            access_token=access_token,
            status='connected',
            last_sync=datetime.utcnow()
        )
        db.add(connection)

    db.commit()
    return {"status": "connected", "message": "Banco conectado exitosamente"}


@router.post("/plaid/sync")
async def sync_plaid_transactions(
    background_tasks: BackgroundTasks,
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Sincronizar transacciones desde Plaid"""
    connection = db.query(Connection).filter_by(
        user_id=user.id,
        provider='plaid'
    ).first()

    if not connection:
        raise HTTPException(status_code=400, detail="Plaid no está conectado")

    # Obtener transacciones
    transactions = await plaid_client.get_transactions(connection.access_token, days=90)

    # Convertir formato Plaid al formato esperado
    items = []
    for txn in transactions:
        item = {
            "id": txn.get("transaction_id"),
            "date": txn.get("date"),
            "amount": abs(float(txn.get("amount", 0))),
            "merchant": txn.get("name", "Transacción"),
            "category": txn.get("personal_finance_category", {}).get("primary", "Otros") if txn.get("personal_finance_category") else "Otros",
            "tx_type": "income" if float(txn.get("amount", 0)) > 0 else "expense",
            "source": "plaid",
            "payment_method": "Transferencia",
            "currency": "ARS",
            "raw_reference": txn.get("transaction_id", ""),
        }
        items.append(item)

    # Guardar transacciones
    saved = await _save_transactions(items, user_id=user.id, db=db)

    connection.last_sync = datetime.utcnow()
    db.commit()

    # Ejecutar deduplicación y alertas
    flagged = mark_duplicates_and_transfers(db, user.id)
    splits = detect_split_candidates(db, user.id)
    background_tasks.add_task(run_alert_engine, user.id, db)

    return {
        "ok": True,
        "fetched": len(transactions),
        "saved": saved,
        "flagged": flagged,
        "split_suggestions": splits
    }


# === PROMETEO ENDPOINTS ===

@router.get("/prometeo/providers")
async def list_prometeo_providers(user: User = Depends(get_current_user)):
    """Listar bancos disponibles en Prometeo"""
    client = get_prometeo_client()
    if not client:
        raise HTTPException(status_code=400, detail="Prometeo no configurado")
    providers = await client.list_providers()
    return {"providers": providers}


class PrometeoLoginRequest(BaseModel):
    provider: str
    username: str
    password: str
    doc_type: str = "C"


@router.post("/prometeo/login")
async def prometeo_login(
    body: PrometeoLoginRequest,
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Login a un banco vía Prometeo con credenciales del usuario"""
    client = get_prometeo_client()
    if not client:
        raise HTTPException(status_code=400, detail="Prometeo no configurado")

    result = await client.login(body.provider, body.username, body.password, body.doc_type)
    if not result or result.get("status") == "error":
        raise HTTPException(status_code=400, detail=result.get("message", "Error conectando al banco"))

    status = result.get("status")
    session_key = result.get("session_key")

    if status == "wrong_credentials":
        raise HTTPException(status_code=401, detail="Usuario o contraseña incorrectos")

    if status not in ("logged_in", "select_client"):
        raise HTTPException(status_code=400, detail=f"Estado inesperado: {status}. {result.get('message','')}")

    # Guardar sesión en BD
    conn = db.query(Connection).filter_by(user_id=user.id, provider="prometeo").first()
    if conn:
        conn.access_token = session_key
        conn.status = "connected"
        conn.last_sync = datetime.utcnow()
    else:
        conn = Connection(
            user_id=user.id,
            provider="prometeo",
            access_token=session_key,
            status="connected",
            last_sync=datetime.utcnow(),
        )
        db.add(conn)
    db.commit()

    return {"status": "connected", "session_key": session_key, "bank_status": status}


@router.post("/prometeo/sync")
async def sync_prometeo_transactions(
    background_tasks: BackgroundTasks,
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Sincronizar transacciones desde Prometeo usando session key guardada"""
    client = get_prometeo_client()
    if not client:
        raise HTTPException(status_code=400, detail="Prometeo no configurado")

    conn = db.query(Connection).filter_by(user_id=user.id, provider="prometeo").first()
    if not conn or conn.status != "connected":
        raise HTTPException(status_code=400, detail="Prometeo no conectado. Conectá tu banco primero.")

    session_key = conn.access_token

    # Obtener cuentas
    accounts = await client.get_accounts(session_key)
    if not accounts:
        raise HTTPException(status_code=400, detail="No se pudieron obtener las cuentas. La sesión puede haber expirado.")

    # Obtener movimientos de todas las cuentas
    all_txns = []
    for acc in accounts:
        number = acc.get("number") or acc.get("account_id", "")
        currency = acc.get("currency", "ARS")
        if number:
            movs = await client.get_movements(session_key, number, currency, days=90)
            all_txns.extend(movs)

    # Convertir al formato interno
    items = []
    for txn in all_txns:
        amount = float(txn.get("amount", 0))
        items.append({
            "id": txn.get("transaction_id"),
            "date": txn.get("date"),
            "amount": abs(amount),
            "merchant": txn.get("name", "Transacción"),
            "tx_type": "income" if amount < 0 else "expense",
            "source": "prometeo",
            "payment_method": "Transferencia Bancaria",
            "currency": txn.get("iso_currency_code", "ARS"),
            "raw_reference": txn.get("transaction_id", ""),
        })

    saved = await _save_transactions(items, user_id=user.id, db=db)
    conn.last_sync = datetime.utcnow()
    db.commit()

    flagged = mark_duplicates_and_transfers(db, user.id)
    splits = detect_split_candidates(db, user.id)
    background_tasks.add_task(run_alert_engine, user.id, db)

    return {"ok": True, "fetched": len(all_txns), "saved": saved, "accounts": len(accounts), "flagged": flagged, "split_suggestions": splits}

File: services/transactions-service/src/routes/sync.js
/** Sync endpoints — Gmail, Mercado Pago, Plaid, y Prometeo. */
import { Router } from "express";
import multer from "multer";
import { z } from "zod";
import { prisma } from "@guita-coach/db";
import { requireAuth, ah, HttpError, detectSplitCandidates, encrypt, decrypt, gmailSvc, mpSvc } from "@guita-coach/shared";
import { findCrossSourceDuplicate, markDuplicatesAndTransfers } from "../services/dedup.js";
import { classify } from "../services/classifier.js";
import { runAlertEngine } from "../services/alertEngine.js";
import { plaidClient } from "../services/plaidSync.js";
import { getPrometeoClient } from "../services/prometeoApi.js";
import { parseMpCsv } from "../services/csvImport.js";
export const syncRouter = Router();
syncRouter.use(requireAuth);
const upload = multer({
  limits: {
    fileSize: 5_000_000
  }
});
async function saveTransactionItem(item, userId) {
  const rawRef = String(item.id || item.raw_reference || "");
  const source = item.source || "";
  if (rawRef) {
    const exists = await prisma.transaction.findFirst({
      where: {
        userId,
        rawReference: rawRef,
        source
      }
    });
    if (exists) return false;
  }
  const dup = await findCrossSourceDuplicate(userId, item);
  const txType = item.tx_type || "expense";
  const paymentMethod = item.payment_method || "";
  if (txType === "income") {
    await prisma.transaction.create({
      data: {
        userId,
        source,
        txType: "income",
        amount: item.amount,
        currency: item.currency || "ARS",
        date: item.date,
        month: item.month || item.date.slice(0, 7),
        merchant: item.merchant || "",
        provider: item.provider || "",
        category: "ingreso",
        subcategory: "",
        status: "confirmed",
        confidence: 1.0,
        paymentMethod,
        needsReview: item.needs_review || false,
        rawReference: rawRef,
        isDuplicate: dup !== null
      }
    });
  } else {
    const result = await classify(item.merchant || "", item.amount || 0, source, userId);
    await prisma.transaction.create({
      data: {
        userId,
        source,
        txType: "expense",
        amount: item.amount,
        currency: item.currency || "ARS",
        date: item.date,
        month: item.month || item.date.slice(0, 7),
        merchant: item.merchant || "",
        provider: item.provider || "",
        category: result.category || "",
        subcategory: result.subcategory || "",
        status: "confirmed",
        confidence: result.confidence ?? 0.7,
        ruleUsed: result.rule_used,
        aiReason: result.ai_reason,
        paymentMethod,
        needsReview: item.needs_review || result.needs_review || false,
        rawReference: rawRef,
        isDuplicate: dup !== null
      }
    });
  }
  return true;
}
async function saveTransactions(items, userId) {
  let saved = 0;
  for (const item of items) {
    if (await saveTransactionItem(item, userId)) saved += 1;
  }
  return saved;
}
syncRouter.post("/gmail", ah(async (req, res) => {
  const user = req.user;
  const conn = await prisma.connection.findFirst({
    where: {
      userId: user.id,
      provider: "gmail"
    }
  });
  if (!conn || conn.status !== "connected" || !conn.accessToken) {
    throw new HttpError(400, "Gmail no conectado. Conectalo desde Configuración.");
  }
  let items;
  try {
    items = await gmailSvc.fetchPaymentEmails(decrypt(conn.accessToken));
  } catch (e) {
    throw new HttpError(502, `Error al leer Gmail: ${e instanceof Error ? e.message : String(e)}`);
  }
  const saved = await saveTransactions(items, user.id);
  await prisma.connection.update({
    where: {
      id: conn.id
    },
    data: {
      lastSync: new Date()
    }
  });
  const flagged = await markDuplicatesAndTransfers(user.id);
  const splits = await detectSplitCandidates(user.id);
  void runAlertEngine(user.id);
  res.json({
    ok: true,
    fetched: items.length,
    saved,
    flagged,
    split_suggestions: splits
  });
}));
syncRouter.post("/mp", ah(async (req, res) => {
  const user = req.user;
  const conn = await prisma.connection.findFirst({
    where: {
      userId: user.id,
      provider: "mercadopago"
    }
  });
  if (!conn || conn.status !== "connected" || !conn.accessToken) {
    throw new HttpError(400, "Mercado Pago no conectado. Conectalo desde Configuración.");
  }
  let items;
  try {
    items = await mpSvc.fetchMovements(decrypt(conn.accessToken));
  } catch (e) {
    throw new HttpError(502, `Error al leer Mercado Pago: ${e instanceof Error ? e.message : String(e)}`);
  }
  const saved = await saveTransactions(items, user.id);
  await prisma.connection.update({
    where: {
      id: conn.id
    },
    data: {
      lastSync: new Date()
    }
  });
  const flagged = await markDuplicatesAndTransfers(user.id);
  const splits = await detectSplitCandidates(user.id);
  void runAlertEngine(user.id);
  res.json({
    ok: true,
    fetched: items.length,
    saved,
    flagged,
    split_suggestions: splits
  });
}));
syncRouter.post("/csv", upload.single("file"), ah(async (req, res) => {
  const user = req.user;
  if (!req.file) throw new HttpError(400, "Archivo requerido");
  if (req.file.size > 5_000_000) throw new HttpError(413, "Archivo muy grande (máximo 5 MB)");
  let items;
  try {
    items = parseMpCsv(req.file.buffer);
  } catch (e) {
    throw new HttpError(400, e instanceof Error ? e.message : String(e));
  }
  const saved = await saveTransactions(items, user.id);
  const flagged = await markDuplicatesAndTransfers(user.id);
  const splits = await detectSplitCandidates(user.id);
  void runAlertEngine(user.id);
  res.json({
    ok: true,
    fetched: items.length,
    saved,
    flagged,
    split_suggestions: splits
  });
}));
syncRouter.get("/status", ah(async (req, res) => {
  const conns = await prisma.connection.findMany({
    where: {
      userId: req.user.id
    }
  });
  const out = {};
  for (const c of conns) {
    out[c.provider] = {
      status: c.status,
      last_sync: c.lastSync ? c.lastSync.toISOString() : null
    };
  }
  res.json(out);
}));

// === PLAID ===

syncRouter.post("/plaid/link_token", ah(async (req, res) => {
  const user = req.user;
  const linkToken = await plaidClient.createLinkToken(user.id, user.email || "");
  if (!linkToken) throw new HttpError(400, "Error creando link token de Plaid");
  res.json({
    link_token: linkToken
  });
}));
const ExchangeTokenSchema = z.object({
  public_token: z.string()
});
syncRouter.post("/plaid/exchange_token", ah(async (req, res) => {
  const user = req.user;
  const {
    public_token
  } = ExchangeTokenSchema.parse(req.body);
  const accessToken = await plaidClient.exchangeToken(public_token);
  if (!accessToken) throw new HttpError(400, "Error intercambiando token");
  const connection = await prisma.connection.findFirst({
    where: {
      userId: user.id,
      provider: "plaid"
    }
  });
  if (connection) {
    await prisma.connection.update({
      where: {
        id: connection.id
      },
      data: {
        accessToken: encrypt(accessToken),
        status: "connected",
        lastSync: new Date()
      }
    });
  } else {
    await prisma.connection.create({
      data: {
        userId: user.id,
        provider: "plaid",
        accessToken: encrypt(accessToken),
        status: "connected",
        lastSync: new Date()
      }
    });
  }
  res.json({
    status: "connected",
    message: "Banco conectado exitosamente"
  });
}));
syncRouter.post("/plaid/sync", ah(async (req, res) => {
  const user = req.user;
  const connection = await prisma.connection.findFirst({
    where: {
      userId: user.id,
      provider: "plaid"
    }
  });
  if (!connection) throw new HttpError(400, "Plaid no está conectado");
  const transactions = await plaidClient.getTransactions(decrypt(connection.accessToken), 90);
  const items = transactions.map(txn => ({
    id: txn.transaction_id,
    date: txn.date,
    amount: Math.abs(Number(txn.amount || 0)),
    merchant: txn.name || "Transacción",
    tx_type: Number(txn.amount || 0) > 0 ? "income" : "expense",
    source: "plaid",
    payment_method: "Transferencia",
    currency: "ARS",
    raw_reference: txn.transaction_id || ""
  }));
  const saved = await saveTransactions(items, user.id);
  await prisma.connection.update({
    where: {
      id: connection.id
    },
    data: {
      lastSync: new Date()
    }
  });
  const flagged = await markDuplicatesAndTransfers(user.id);
  const splits = await detectSplitCandidates(user.id);
  void runAlertEngine(user.id);
  res.json({
    ok: true,
    fetched: transactions.length,
    saved,
    flagged,
    split_suggestions: splits
  });
}));

// === PROMETEO ===

syncRouter.get("/prometeo/providers", ah(async (_req, res) => {
  const client = getPrometeoClient();
  if (!client) throw new HttpError(400, "Prometeo no configurado");
  const providers = await client.listProviders();
  res.json({
    providers
  });
}));
const PrometeoLoginSchema = z.object({
  provider: z.string(),
  username: z.string(),
  password: z.string(),
  doc_type: z.string().default("C")
});
syncRouter.post("/prometeo/login", ah(async (req, res) => {
  const user = req.user;
  const client = getPrometeoClient();
  if (!client) throw new HttpError(400, "Prometeo no configurado");
  const body = PrometeoLoginSchema.parse(req.body);
  const result = await client.login(body.provider, body.username, body.password, body.doc_type);
  if (!result || result.status === "error") {
    throw new HttpError(400, result?.message || "Error conectando al banco");
  }
  const status = result.status;
  const sessionKey = result.session_key;
  if (status === "wrong_credentials") throw new HttpError(401, "Usuario o contraseña incorrectos");
  if (status !== "logged_in" && status !== "select_client") {
    throw new HttpError(400, `Estado inesperado: ${status}. ${result.message || ""}`);
  }
  const conn = await prisma.connection.findFirst({
    where: {
      userId: user.id,
      provider: "prometeo"
    }
  });
  if (conn) {
    await prisma.connection.update({
      where: {
        id: conn.id
      },
      data: {
        accessToken: encrypt(sessionKey),
        status: "connected",
        lastSync: new Date()
      }
    });
  } else {
    await prisma.connection.create({
      data: {
        userId: user.id,
        provider: "prometeo",
        accessToken: encrypt(sessionKey),
        status: "connected",
        lastSync: new Date()
      }
    });
  }
  res.json({
    status: "connected",
    session_key: sessionKey,
    bank_status: status
  });
}));
syncRouter.post("/prometeo/sync", ah(async (req, res) => {
  const user = req.user;
  const client = getPrometeoClient();
  if (!client) throw new HttpError(400, "Prometeo no configurado");
  const conn = await prisma.connection.findFirst({
    where: {
      userId: user.id,
      provider: "prometeo"
    }
  });
  if (!conn || conn.status !== "connected") throw new HttpError(400, "Prometeo no conectado. Conectá tu banco primero.");
  const sessionKey = decrypt(conn.accessToken);
  const accounts = await client.getAccounts(sessionKey);
  if (!accounts.length) throw new HttpError(400, "No se pudieron obtener las cuentas. La sesión puede haber expirado.");
  const allTxns = [];
  for (const acc of accounts) {
    const number = acc.number || acc.account_id || "";
    const currency = acc.currency || "ARS";
    if (number) {
      const movs = await client.getMovements(sessionKey, number, currency, 90);
      allTxns.push(...movs);
    }
  }
  const items = allTxns.map(txn => {
    const amount = Number(txn.amount || 0);
    return {
      id: txn.transaction_id,
      date: txn.date,
      amount: Math.abs(amount),
      merchant: txn.name || "Transacción",
      tx_type: amount < 0 ? "income" : "expense",
      source: "prometeo",
      payment_method: "Transferencia Bancaria",
      currency: txn.iso_currency_code || "ARS",
      raw_reference: txn.transaction_id || ""
    };
  });
  const saved = await saveTransactions(items, user.id);
  await prisma.connection.update({
    where: {
      id: conn.id
    },
    data: {
      lastSync: new Date()
    }
  });
  const flagged = await markDuplicatesAndTransfers(user.id);
  const splits = await detectSplitCandidates(user.id);
  void runAlertEngine(user.id);
  res.json({
    ok: true,
    fetched: allTxns.length,
    saved,
    accounts: accounts.length,
    flagged,
    split_suggestions: splits
  });
}));
